this is a note book for discovring source data and mini analysis

In [2]:
%pip install boto3 pandas

Note: you may need to restart the kernel to use updated packages.


In [10]:
import boto3
import pandas as pd

BUCKET = "cct-ds-code-challenge-input-data"
REGION = "af-south-1"

s3 = boto3.client(
    "s3",
    region_name=REGION
)

In [ ]:
response = s3.list_objects_v2(Bucket=BUCKET)

for file in response["Contents"]:
    print(file["Key"])

city-hex-polygons-7.geojson
city-hex-polygons-8-10.geojson
city-hex-polygons-8.geojson
ds_code_challenge_creds.json
images/swimming-pool/no/W07C_11_100.tif
images/swimming-pool/no/W07C_11_61.tif
images/swimming-pool/no/W07C_11_71.tif
images/swimming-pool/no/W07C_11_82.tif
images/swimming-pool/no/W07C_11_88.tif
images/swimming-pool/no/W07C_11_97.tif
images/swimming-pool/no/W07C_16_1.tif
images/swimming-pool/no/W07C_16_10.tif
images/swimming-pool/no/W07C_16_17.tif
images/swimming-pool/no/W07C_16_24.tif
images/swimming-pool/no/W07C_16_25.tif
images/swimming-pool/no/W07C_16_27.tif
images/swimming-pool/no/W07C_16_29.tif
images/swimming-pool/no/W07C_16_35.tif
images/swimming-pool/no/W07C_16_36.tif
images/swimming-pool/no/W07C_16_38.tif
images/swimming-pool/no/W07C_16_39.tif
images/swimming-pool/no/W07C_16_40.tif
images/swimming-pool/no/W07C_16_45.tif
images/swimming-pool/no/W07C_16_46.tif
images/swimming-pool/no/W07C_16_47.tif
images/swimming-pool/no/W07C_16_5.tif
images/swimming-pool/no/W07

In [1]:
import boto3
import pandas as pd

BUCKET = "cct-ds-code-challenge-input-data"
REGION = "af-south-1"

s3 = boto3.client(
    "s3",
    region_name=REGION
)

response = s3.list_objects_v2(Bucket=BUCKET)

files = []

for obj in response.get("Contents", []):
    files.append({
        "key": obj["Key"],
        "size_bytes": obj["Size"],
        "last_modified": obj["LastModified"]
    })

files_df = pd.DataFrame(files)

files_df

,key,size_bytes,last_modified
0,city-hex-polygons-7.geojson,276231,2021-05-14 09:52:18+00:00
1,city-hex-polygons-8-10.geojson,108254980,2021-05-14 09:44:09+00:00
2,city-hex-polygons-8.geojson,1966685,2021-05-14 13:10:00+00:00
3,ds_code_challenge_creds.json,115,2021-05-14 16:34:40+00:00
4,images/swimming-pool/no/W07C_11_100.tif,3932696,2025-06-12 09:03:22+00:00
...,...,...,...
995,images/swimming-pool/no/W17D_22_41.tif,3911391,2025-06-12 09:03:38+00:00
996,images/swimming-pool/no/W17D_22_45.tif,4134574,2025-06-12 09:03:38+00:00
997,images/swimming-pool/no/W17D_22_46.tif,4125109,2025-06-12 09:03:38+00:00
998,images/swimming-pool/no/W17D_22_47.tif,4147367,2025-06-12 09:03:38+00:00


In [3]:
import json


print(json.dumps(response, indent=4, default=str))

{
    "ResponseMetadata": {
        "RequestId": "7AH9WZ1G5PMPVEFK",
        "HostId": "WEqdhvabktRAR/tYrtrF50A8f4J31+yTgSQrhp0ZrMPdZci0e47EyA3++cNOBQP5n4Jhm8q9Yk2UQuUCX8HSgx0Wihn889O7",
        "HTTPStatusCode": 200,
        "HTTPHeaders": {
            "x-amz-id-2": "WEqdhvabktRAR/tYrtrF50A8f4J31+yTgSQrhp0ZrMPdZci0e47EyA3++cNOBQP5n4Jhm8q9Yk2UQuUCX8HSgx0Wihn889O7",
            "x-amz-request-id": "7AH9WZ1G5PMPVEFK",
            "date": "Sat, 22 Aug 2026 08:04:34 GMT",
            "x-amz-bucket-region": "af-south-1",
            "content-type": "application/xml",
            "transfer-encoding": "chunked",
            "server": "AmazonS3"
        },
        "RetryAttempts": 0
    },
    "IsTruncated": true,
    "Contents": [
        {
            "Key": "city-hex-polygons-7.geojson",
            "LastModified": "2021-05-14 09:52:18+00:00",
            "ETag": "\"f04048ba5a5ab4aff096a55935bcd447\"",
            "Size": 276231,
            "StorageClass": "STANDARD"
        },
        {


In [13]:
files_df["file_type"] = (
    files_df["key"]
    .str.lower()
    .str.split(".")
    .str[-1]
)

files_df["file_type"].value_counts()

file_type
tif        996
geojson      3
json         1
Name: count, dtype: int64

In [15]:
import json

KEY = "city-hex-polygons-7.geojson"

response = s3.get_object(
    Bucket=BUCKET,
    Key=KEY
)

geojson_data = json.loads(
    response["Body"].read()
)

type(geojson_data)

dict

In [16]:
geojson_data.keys()

dict_keys(['type', 'name', 'crs', 'features'])

In [17]:
len(geojson_data["features"])

627

In [18]:
geojson_data["features"][0]

{'type': 'Feature',
 'properties': {'index': '87ad36180ffffff'},
 'geometry': {'type': 'Polygon',
  'coordinates': [[[18.69026825870867, -33.86703319751351],
    [18.69168980931549, -33.85511764511594],
    [18.679265404549795, -33.84751122474886],
    [18.66542004486691, -33.8518201940735],
    [18.663997091762614, -33.86373510896562],
    [18.676420900551488, -33.87134169227918],
    [18.69026825870867, -33.86703319751351]]]}}

In [19]:
features_df = pd.json_normalize(
    geojson_data["features"]
)

features_df.head()

,type,properties.index,geometry.type,geometry.coordinates
0,Feature,87ad36180ffffff,Polygon,"[[[18.69026825870867, -33.86703319751351], [18..."
1,Feature,87ad36181ffffff,Polygon,"[[[18.663997091762614, -33.86373510896562], [1..."
2,Feature,87ad36182ffffff,Polygon,"[[[18.705536812042357, -33.850805737933875], [..."
3,Feature,87ad36183ffffff,Polygon,"[[[18.679265404549795, -33.84751122474886], [1..."
4,Feature,87ad36184ffffff,Polygon,"[[[18.70127401694917, -33.88655041755364], [18..."


In [20]:
# Number of rows and columns
features_df.shape

(627, 4)

In [21]:
features_df.dtypes

type                    object
properties.index        object
geometry.type           object
geometry.coordinates    object
dtype: object

In [22]:
# Check whether the hex IDs are unique
features_df["properties.index"].nunique()

627

discovery finding:

Source: city-hex-polygons-7.geojson
Format: GeoJSON
Records: 627
Geometry: Polygon
Business/geospatial identifier: properties.index
Unique identifiers: 627 / 627

In [23]:
features_df.isnull().sum()

type                    0
properties.index        0
geometry.type           0
geometry.coordinates    0
dtype: int64

In [24]:
features_df["geometry.type"].value_counts()

geometry.type
Polygon    627
Name: count, dtype: int64

In [25]:
features_df["type"].value_counts()

type
Feature    627
Name: count, dtype: int64

what axctly is in the the 1,000 objects 

In [26]:
files_df["file_type"].value_counts()

file_type
tif        996
geojson      3
json         1
Name: count, dtype: int64

In [27]:
files_df[
    files_df["key"].str.startswith("images/")
]["key"].head(20)

4     images/swimming-pool/no/W07C_11_100.tif
5      images/swimming-pool/no/W07C_11_61.tif
6      images/swimming-pool/no/W07C_11_71.tif
7      images/swimming-pool/no/W07C_11_82.tif
8      images/swimming-pool/no/W07C_11_88.tif
9      images/swimming-pool/no/W07C_11_97.tif
10      images/swimming-pool/no/W07C_16_1.tif
11     images/swimming-pool/no/W07C_16_10.tif
12     images/swimming-pool/no/W07C_16_17.tif
13     images/swimming-pool/no/W07C_16_24.tif
14     images/swimming-pool/no/W07C_16_25.tif
15     images/swimming-pool/no/W07C_16_27.tif
16     images/swimming-pool/no/W07C_16_29.tif
17     images/swimming-pool/no/W07C_16_35.tif
18     images/swimming-pool/no/W07C_16_36.tif
19     images/swimming-pool/no/W07C_16_38.tif
20     images/swimming-pool/no/W07C_16_39.tif
21     images/swimming-pool/no/W07C_16_40.tif
22     images/swimming-pool/no/W07C_16_45.tif
23     images/swimming-pool/no/W07C_16_46.tif
Name: key, dtype: object

In [29]:
service_request_files = files_df[
    files_df["key"].str.contains(
        r"(^|/)(sr|sr_hex|sr_hex_truncated)",
        case=False,
        regex=True
    )
]

service_request_files

C:\Users\User\AppData\Local\Temp\ipykernel_17352\290442126.py:2: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  files_df["key"].str.contains(


,key,size_bytes,last_modified,file_type


In [30]:
print(f"Total S3 objects discovered: {len(files_df):,}")

Total S3 objects discovered: 1,000


In [32]:
geojson_files = files_df[
    files_df["key"].str.contains(
        "city-hex",
        case=False,
        na=False
    )
]

geojson_files

,key,size_bytes,last_modified,file_type
0,city-hex-polygons-7.geojson,276231,2021-05-14 09:52:18+00:00,geojson
1,city-hex-polygons-8-10.geojson,108254980,2021-05-14 09:44:09+00:00,geojson
2,city-hex-polygons-8.geojson,1966685,2021-05-14 13:10:00+00:00,geojson


In [33]:
required_sources = [
    "sr.csv.gz",
    "sr_hex.csv.gz",
    "sr_hex_truncated.csv",
    "city-hex-polygons-8.geojson",
    "city-hex-polygons-8-10.geojson",
]

source_check = []

for key in required_sources:
    try:
        response = s3.head_object(
            Bucket=BUCKET,
            Key=key
        )

        source_check.append({
            "key": key,
            "exists": True,
            "size_bytes": response["ContentLength"],
            "last_modified": response["LastModified"],
        })

    except s3.exceptions.ClientError as exc:
        source_check.append({
            "key": key,
            "exists": False,
            "size_bytes": None,
            "last_modified": None,
            "error": exc.response["Error"]["Code"],
        })

source_check_df = pd.DataFrame(source_check)

source_check_df

,key,exists,size_bytes,last_modified
0,sr.csv.gz,True,37183035,2022-06-21 20:13:07+00:00
1,sr_hex.csv.gz,True,36657416,2022-06-21 20:13:19+00:00
2,sr_hex_truncated.csv,True,63403673,2022-06-21 20:13:32+00:00
3,city-hex-polygons-8.geojson,True,1966685,2021-05-14 13:10:00+00:00
4,city-hex-polygons-8-10.geojson,True,108254980,2021-05-14 09:44:09+00:00


one we have pictur of our sources in the s3 bucket as specified

In [34]:
required_sources = [
    "city-hex-polygons-8-10.geojson",
    "city-hex-polygons-8.geojson",
    "sr.csv.gz",
    "sr_hex.csv.gz",
    "sr_hex_truncated.csv",
]

source_inventory = files_df[
    files_df["key"].isin(required_sources)
].copy()

source_inventory

,key,size_bytes,last_modified,file_type
1,city-hex-polygons-8-10.geojson,108254980,2021-05-14 09:44:09+00:00,geojson
2,city-hex-polygons-8.geojson,1966685,2021-05-14 13:10:00+00:00,geojson


get all the object in the storage

In [35]:
required_sources = [
    "city-hex-polygons-8-10.geojson",
    "city-hex-polygons-8.geojson",
    "sr.csv.gz",
    "sr_hex.csv.gz",
    "sr_hex_truncated.csv",
]

source_inventory = []

for key in required_sources:
    response = s3.head_object(
        Bucket=BUCKET,
        Key=key
    )

    source_inventory.append({
        "key": key,
        "size_bytes": response["ContentLength"],
        "last_modified": response["LastModified"],
    })

source_inventory_df = pd.DataFrame(source_inventory)

source_inventory_df

,key,size_bytes,last_modified
0,city-hex-polygons-8-10.geojson,108254980,2021-05-14 09:44:09+00:00
1,city-hex-polygons-8.geojson,1966685,2021-05-14 13:10:00+00:00
2,sr.csv.gz,37183035,2022-06-21 20:13:07+00:00
3,sr_hex.csv.gz,36657416,2022-06-21 20:13:19+00:00
4,sr_hex_truncated.csv,63403673,2022-06-21 20:13:32+00:00


In [37]:
%pip install s3fs

  Using cached async_timeout-5.0.1-py3-none-any.whl (6.2 kB)
  Attempting uninstall: botocore
    Found existing installation: botocore 1.43.77
    Uninstalling botocore-1.43.77:
      Successfully uninstalled botocore-1.43.77
Note: you may need to restart the kernel to use updated packages.


ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
boto3 1.43.77 requires botocore<1.44.0,>=1.43.77, but you have botocore 1.43.56 which is incompatible.


In [38]:
import pandas as pd

sr = pd.read_csv(
    f"s3://{BUCKET}/sr.csv.gz",
    compression="gzip"
)

print(f"Rows: {len(sr):,}")
print(f"Columns: {len(sr.columns):,}")

Rows: 941,634
Columns: 16


In [39]:
sr.columns.tolist()

['Unnamed: 0',
 'notification_number',
 'reference_number',
 'creation_timestamp',
 'completion_timestamp',
 'directorate',
 'department',
 'branch',
 'section',
 'code_group',
 'code',
 'cause_code_group',
 'cause_code',
 'official_suburb',
 'latitude',
 'longitude']

In [40]:
sr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 941634 entries, 0 to 941633
Data columns (total 16 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Unnamed: 0            941634 non-null  int64  
 1   notification_number   941634 non-null  int64  
 2   reference_number      592920 non-null  float64
 3   creation_timestamp    941634 non-null  object 
 4   completion_timestamp  929442 non-null  object 
 5   directorate           932199 non-null  object 
 6   department            932180 non-null  object 
 7   branch                913233 non-null  object 
 8   section               848509 non-null  object 
 9   code_group            941634 non-null  object 
 10  code                  941634 non-null  object 
 11  cause_code_group      131117 non-null  object 
 12  cause_code            129669 non-null  object 
 13  official_suburb       729221 non-null  object 
 14  latitude              729270 non-null  float64
 15  

In [41]:
sr.head()

,Unnamed: 0,notification_number,reference_number,creation_timestamp,completion_timestamp,directorate,department,branch,section,code_group,code,cause_code_group,cause_code,official_suburb,latitude,longitude
0,0,400583534,9.109492e+09,2020-10-07 06:55:18+02:00,2020-10-08 15:36:35+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area Central,District: Blaauwberg,TD Customer complaint groups,Pothole&Defect Road Foot Bic Way/Kerbs,Road (RCL),Wear and tear,MONTAGUE GARDENS,-33.872839,18.522488
1,1,400555043,9.108995e+09,2020-07-09 16:08:13+02:00,2020-07-14 14:27:01+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area East,District : Somerset West,TD Customer complaint groups,Manhole Cover/Gully Grid,Road (RCL),Vandalism,SOMERSET WEST,-34.078916,18.848940
2,2,400589145,9.109614e+09,2020-10-27 10:21:59+02:00,2020-10-28 17:48:15+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area East,District : Somerset West,TD Customer complaint groups,Manhole Cover/Gully Grid,Road (RCL),Vandalism,STRAND,-34.102242,18.821116
3,3,400538915,9.108601e+09,2020-03-19 06:36:06+02:00,2021-03-29 20:34:19+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area North,District : Bellville,TD Customer complaint groups,Paint Markings Lines&Signs,Road Markings,Wear and tear,RAVENSMEAD,-33.920019,18.607209
4,4,400568554,NaN,2020-08-25 09:48:42+02:00,2020-08-31 08:41:13+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area South,District : Athlone,TD Customer complaint groups,Pothole&Defect Road Foot Bic Way/Kerbs,Road (RCL),Surfacing failure,CLAREMONT,-33.987400,18.453760


In [45]:
sr[["latitude", "longitude"]].isna().sum()

latitude     212364
longitude    212364
dtype: int64

In [46]:
coordinate_nulls = sr[["latitude", "longitude"]].isna().sum()

coordinate_nulls

latitude     212364
longitude    212364
dtype: int64

In [47]:
print("Total records:", len(sr))
print("Records with coordinates:", sr[["latitude", "longitude"]].notna().all(axis=1).sum())
print("Records missing coordinates:", sr[["latitude", "longitude"]].isna().any(axis=1).sum())
print("Coordinate completeness:", f"{sr[['latitude', 'longitude']].notna().all(axis=1).mean():.2%}")

Total records: 941634
Records with coordinates: 729270
Records missing coordinates: 212364
Coordinate completeness: 77.45%


Load sr_hex.csv.gz

In [48]:
sr_hex =pd.read_csv(f"s3://{BUCKET}/sr_hex.csv.gz", compression="gzip")

In [49]:
print("Rows:", len(sr_hex))
print("Columns:", len(sr_hex.columns))

Rows: 941634
Columns: 16


In [56]:
sr_hex.columns.tolist()

['notification_number',
 'reference_number',
 'creation_timestamp',
 'completion_timestamp',
 'directorate',
 'department',
 'branch',
 'section',
 'code_group',
 'code',
 'cause_code_group',
 'cause_code',
 'official_suburb',
 'latitude',
 'longitude',
 'h3_level8_index']

In [57]:
sr_hex.head(10)

,notification_number,reference_number,creation_timestamp,completion_timestamp,directorate,department,branch,section,code_group,code,cause_code_group,cause_code,official_suburb,latitude,longitude,h3_level8_index
0,400583534,9.109492e+09,2020-10-07 06:55:18+02:00,2020-10-08 15:36:35+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area Central,District: Blaauwberg,TD Customer complaint groups,Pothole&Defect Road Foot Bic Way/Kerbs,Road (RCL),Wear and tear,MONTAGUE GARDENS,-33.872839,18.522488,88ad360225fffff
1,400555043,9.108995e+09,2020-07-09 16:08:13+02:00,2020-07-14 14:27:01+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area East,District : Somerset West,TD Customer complaint groups,Manhole Cover/Gully Grid,Road (RCL),Vandalism,SOMERSET WEST,-34.078916,18.848940,88ad36d5e1fffff
2,400589145,9.109614e+09,2020-10-27 10:21:59+02:00,2020-10-28 17:48:15+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area East,District : Somerset West,TD Customer complaint groups,Manhole Cover/Gully Grid,Road (RCL),Vandalism,STRAND,-34.102242,18.821116,88ad36d437fffff
3,400538915,9.108601e+09,2020-03-19 06:36:06+02:00,2021-03-29 20:34:19+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area North,District : Bellville,TD Customer complaint groups,Paint Markings Lines&Signs,Road Markings,Wear and tear,RAVENSMEAD,-33.920019,18.607209,88ad361133fffff
4,400568554,NaN,2020-08-25 09:48:42+02:00,2020-08-31 08:41:13+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area South,District : Athlone,TD Customer complaint groups,Pothole&Defect Road Foot Bic Way/Kerbs,Road (RCL),Surfacing failure,CLAREMONT,-33.987400,18.453760,88ad361709fffff
5,400556029,9.109003e+09,2020-07-11 10:34:37+02:00,2020-07-21 12:35:53+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area East,District : Somerset West,TD Customer complaint groups,Pothole&Defect Road Foot Bic Way/Kerbs,Road (RCL),Base failure,STRAND,-34.119520,18.829917,88ad36d43dfffff
6,400588369,NaN,2020-10-23 10:33:48+02:00,2020-10-26 14:16:49+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area North,District : Bellville,TD Customer complaint groups,Pothole&Defect Road Foot Bic Way/Kerbs,NaN,NaN,GLENHAVEN,-33.917996,18.658031,88ad361a19fffff
7,400549080,NaN,2020-06-17 15:46:11+02:00,2020-11-24 14:26:09+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area North,District : Bellville,TD Customer complaint groups,Pothole&Defect Road Foot Bic Way/Kerbs,NaN,NaN,RAVENSMEAD,-33.922154,18.596821,88ad3611edfffff
8,400599701,9.109845e+09,2020-12-04 08:30:50+02:00,2020-12-14 12:53:52+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area Central,District: Cape Town,TD Customer complaint groups,Block Pipe/Chan/Canal/Drain/Gully/Manh,NaN,NaN,CAPE TOWN CITY CENTRE,-33.925721,18.421675,88ad361519fffff
9,400569986,NaN,2020-08-28 13:37:53+02:00,2020-08-28 16:17:35+02:00,URBAN MOBILITY,Roads Infrastructure Management,RIM Area North,District : Kraaifontein,TD Customer complaint groups,Pothole&Defect Road Foot Bic Way/Kerbs,Road (RCL),Surfacing failure,DURBANVILLE,-33.835098,18.652708,88ad3618e5fffff


In [58]:
coordinate_nulls = sr_hex[["latitude", "longitude"]].isna().sum()

In [59]:
coordinate_nulls

latitude     212364
longitude    212364
dtype: int64

Compare row counts

In [51]:
set(sr_hex.columns) - set(sr.columns)

{'h3_level8_index'}

In [52]:
print("sr rows:", len(sr))
print("sr_hex rows:", len(sr_hex))

sr rows: 941634
sr_hex rows: 941634


In [53]:
sr_hex["h3_level8_index"].value_counts(dropna=False).head()

h3_level8_index
0                  212364
88ad3612a9fffff      3815
88ad361009fffff      3464
88ad368985fffff      3393
88ad361519fffff      2797
Name: count, dtype: int64

In [60]:
print(
    "H3 index 0:",
    (sr_hex["h3_level8_index"] == 0).sum()
)

H3 index 0: 0


In [61]:
print(
    "H3 index 0:",
    (sr_hex["h3_level8_index"] == "0").sum()
)

H3 index 0: 212364


In [62]:
print("Total records:", len(sr_hex))
print("H3 index 0:", (sr_hex["h3_level8_index"] == "0").sum())
print("Valid H3 index:", (sr_hex["h3_level8_index"] != "0").sum())

Total records: 941634
H3 index 0: 212364
Valid H3 index: 729270


GeoJSON reference file.

In [63]:
%pip install geopandas

Note: you may need to restart the kernel to use updated packages.


In [64]:
import geopandas as gpd

hex8 = gpd.read_file(
    f"s3://{BUCKET}/city-hex-polygons-8.geojson"
)

print("Rows:", len(hex8))
print("Columns:", len(hex8.columns))

Rows: 3832
Columns: 4


In [65]:
hex8.columns.tolist()

['index', 'centroid_lat', 'centroid_lon', 'geometry']

In [66]:
hex8.head()

,index,centroid_lat,centroid_lon,geometry
0,88ad361801fffff,-33.859427,18.677843,"POLYGON ((18.68119 -33.8633, 18.68357 -33.8592..."
1,88ad361803fffff,-33.855696,18.668766,"POLYGON ((18.67211 -33.85957, 18.6745 -33.8555..."
2,88ad361805fffff,-33.855263,18.685959,"POLYGON ((18.68931 -33.85914, 18.69169 -33.855..."
3,88ad361807fffff,-33.851532,18.676881,"POLYGON ((18.68023 -33.85541, 18.68261 -33.851..."
4,88ad361809fffff,-33.867322,18.678806,"POLYGON ((18.68215 -33.8712, 18.68454 -33.8671..."


In [67]:
print("Hexagons:", len(hex8))
print("Unique H3 indexes:", hex8["index"].nunique())
print("Missing H3 indexes:", hex8["index"].isna().sum())
print("Missing geometry:", hex8["geometry"].isna().sum())

Hexagons: 3832
Unique H3 indexes: 3832
Missing H3 indexes: 0
Missing geometry: 0


In [68]:
print("CRS:", hex8.crs)
print("Geometry types:")
print(hex8.geometry.geom_type.value_counts())

CRS: EPSG:4326
Geometry types:
Polygon    3832
Name: count, dtype: int64


Notes 
Hexagons:              3,832
Unique H3 indexes:     3,832
Missing H3 indexes:    0
Missing geometry:      0
CRS:                   EPSG:4326
Geometry type:         Polygon

we need tocheck if H3 resolution has json data lets checkit

In [69]:
response = s3.get_object(
    Bucket=BUCKET,
    Key="city-hex-polygons-8-10.geojson",
    Range="bytes=0-5000"
)

sample = response["Body"].read().decode("utf-8")

print(sample[:5000])

{
"type": "FeatureCollection",
"crs": { "type": "name", "properties": { "name": "urn:ogc:def:crs:OGC:1.3:CRS84" } },
"features": [
{ "type": "Feature", "properties": { "index": "88ad361801fffff", "centroid_lat": -33.859427322761434, "centroid_lon": 18.677843311941835, "resolution": 8 }, "geometry": { "type": "Polygon", "coordinates": [ [ [ 18.6811898997334, -33.863302790817968 ], [ 18.683574296194426, -33.85928287732969 ], [ 18.680227609989728, -33.85540739558428 ], [ 18.674496767706248, -33.855551865779326 ], [ 18.672112346191998, -33.859571723609463 ], [ 18.675458791982372, -33.863447166906802 ], [ 18.6811898997334, -33.863302790817968 ] ] ] } },
{ "type": "Feature", "properties": { "index": "88ad361803fffff", "centroid_lat": -33.855695953283139, "centroid_lon": 18.668766173796456, "resolution": 8 }, "geometry": { "type": "Polygon", "coordinates": [ [ [ 18.672112346191998, -33.859571723609463 ], [ 18.674496767706248, -33.855551865779326 ], [ 18.671150496788311, -33.851676081758441 ],